Add Partial Content Delivery (+tests)

agersant opened the pull request:
This is a follow up to https://github.com/iron/staticfile/pull/95. @Cobrand doesn't seem to be active on Github anymore but his pending PR adding partial content responses is very valuable to me. I added unit tests to his code, please let me know if anything else needs to happen for this to be merged!

Cobrand committed:
Add Partial Content Delivery

This commits adds Partial Content Delivery, as asked in issue #93. This
enables the "Accept-Ranges" header on all files delivered via
static. It is possible to ask only a certain portion of bytes like
explicitely described in RFC 7233. Of the 3 ways to ask for a range of
bytes (byte x to byte y, everything from byte x to end of file, last y
bytes from end of file), every one of them is implemented, but if a
request is to ask multi ranges in a same request, this implementation
will only account the first range.

agersant committed:
Adds unit test for partial content delivery

Cobrand commented:
I am still active on  GitHub and I saw your PR on my fork, and I wanted to get to it this week-end but you beat me to it. It's true that lately I haven't as much time as I wished I had, because you know, life'n'stuff.

Anyway I'm grateful that you're taking care of this now, I was starting to feel guilty about not implementing the tests to have this merged.

About the tests themselves, if I were you I would also add some tests for the errors as well. (requesting a byterange were the minimum is higher than the file total, requesting a from -1 in the headers, things like that... but still, great work and thank you again.

agersant commented:
I should have been a bit more patient, I know how busy things can get :c

I'll add a few tests for invalid requests, that's a good point.

agersant committed:
Adds unit tests for partial content delivery edge cases

Also fixed a edge case bug where a FromTo range ending out of bounds was not treated as a From range (See https://tools.ietf.org/html/rfc7233, section 2.1).

Hoverbear commented on src/partial_file.rs:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    /// Panics if the file doesn't exist
+    pub fn from_path<P: AsRef<Path>, Range>(path: P, range: Range) -> PartialFile
```
Is there a specific reason this doesn't return a `Result<PartialFile, Std::io::Error>`?

Cobrand replied:
The reason was originally that the file existence was already checked beforehand everytime before this method was called (via metadata). However now that I think of it this is pretty racy: if the file is deleted between the check and the file being delivered (which is unlikely but can still happen), then it will panic.

It would be better to return a `Result<_, io::Error>` there indeed.

agersant committed:
Fixed a bug where I/O errors weren't handled gracefully

agersant commented:
Any update regarding the possibility/timeline of this being merged?

Emielvanbetsbrugge commented:
Hey guys, this works great, we're using this pull request in a video renderer. Should be good to go in. Cheers.

rockwotj commented:
Ping!

rockwotj commented:
Maybe @untitaker would be willing to review/merge this?

agersant commented:
At this point I would recommend migrating to a web framework that is still maintained.

rockwotj commented:
Is this not maintained? A deprecation warning would be nice

agersant commented:
Yeah, Iron hasn't had a commit in a very long time. I think everyone uses Rocket, Actix-Web or some other frameworks now.
